go/client/cmd_simplefs_copy.go

```diff
@@ -0,0 +1,143 @@
+// Copyright 2015 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+package client
+
+import (
+	"errors"
+	"os"
+	"time"
+
+	"golang.org/x/net/context"
+
+	"encoding/hex"
+
+	"github.com/keybase/cli"
+	"github.com/keybase/client/go/libcmdline"
+	"github.com/keybase/client/go/libkb"
+	keybase1 "github.com/keybase/client/go/protocol/keybase1"
+)
+
+// CmdSimpleFSCopy is the 'simplefs list' command.
+type CmdSimpleFSCopy struct {
+	libkb.Contextified
+	opid    keybase1.OpID
+	src     keybase1.Path
+	dest    keybase1.Path
+	recurse bool
+	argOpid bool // set when -o is used
+}
+
+// NewCmdSimpleFSCopy creates a new cli.Command.
+func NewCmdSimpleFSCopy(cl *libcmdline.CommandLine, g *libkb.GlobalContext) cli.Command {
+	return cli.Command{
+		Name:         "cp",
+		ArgumentHelp: "<source> [dest]",
+		Usage:        "copy directory elements",
+		Action: func(c *cli.Context) {
+			cl.ChooseCommand(&CmdDeviceList{Contextified: libkb.NewContextified(g)}, "cp", c)
+		},
+		Flags: []cli.Flag{
+			cli.BoolFlag{
+				Name:  "r, recursive",
+				Usage: "Recurse into subdirectories",
+			},
+			cli.StringFlag{
+				Name:  "o, opid",
+				Usage: "retrieve results",
+			},
+		},
+	}
+}
+
+// RunClient runs the command in client/server mode.
+func (c *CmdSimpleFSCopy) Run() error {
+	cli, err := GetSimpleFSClient(c.G())
+	if err != nil {
+		return err
+	}
+
+	if !c.argOpid {
+		c.opid, err = cli.SimpleFSMakeOpid(context.TODO())
+		defer cli.SimpleFSClose(context.TODO(), c.opid)
+		if err != nil {
+			return err
+		}
+		if c.recurse {
+			err = cli.SimpleFSCopyRecursive(context.TODO(), keybase1.SimpleFSCopyRecursiveArg{
+				OpID: c.opid,
+				Src:  c.src,
+				Dest: c.src,
+			})
+		} else {
+			err = cli.SimpleFSCopy(context.TODO(), keybase1.SimpleFSCopyArg{
+				OpID: c.opid,
+				Dest: c.src,
+			})
+		}
+		if err != nil {
+			return err
+		}
+	}
+
+	for {
+		progress, err := cli.SimpleFSCheck(context.TODO(), c.opid)
+		if err != nil {
+			break
+		}
+		// break if we're done or the opid was provided
+		if progress == 100 || c.argOpid {
```

Comment:
The notion of `progress` ought to be strictly informational.  Correctness shouldn't depend upon it.